### ISSUE TITLE
qemu: secure trace outputs to non-secure console

### ISSUE THREAD

**Issue by etienne-lms** [OP-TEE/optee_os#2848]
In OP-TEE 3.4.0, platform `qemu_virt` outputs the OP-TEE core and TA traces in the Linux console. Shell command `make run` (and friends) opens 2 console windows but the one entitled "Secure" outputs only early OP-TEE traces then remains silent.

The issue is due to the console being probed from the external DT which provides only the non-secure console information.

#### Console init sequence:
At core entry early console initialization, `qemu_virt` registers a TEE console on UART1.
At core initialization `generic_boot.c` reads the DT to get the `stdout-path` from the `secure-chosen` or `chosen` node.

DTB content has no `secure-chosen` node hence core finds the non-secure `chosen` node and its `stdout-path` property, defining the Linux console interface, here UART0. The last traces on secure console are:
```
D/TC:0 0 dt_add_psci_node:639 PSCI Device Tree node already exists!
I/TC: Switching console to device: 
```
and non-secure console starts with:
```
I/TC: OP-TEE version: 3.4.0-106-g8b09a36-dev #6 Mon Feb 25 21:43:41 UTC 2019 arm
D/TC:0 0 check_ta_store:534 TA store: "Secure Storage TA"
D/TC:0 0 check_ta_store:534 TA store: "REE [buffered]"
D/TC:0 0 mobj_mapped_shm_init:721 Shared memory address range: e300000, 10300000
D/TC:0 0 tee_pager_add_core_area:536 0xe184000 - 0xe18f000 : flags 0x30, store 0x0, hashes 0x0
D/TC:0 0 gic_it_set_cpu_mask:251 cpu_mask: writing 0xff to 0x11f00828
D/TC:0 0 gic_it_set_cpu_mask:253 cpu_mask: 0xff
D/TC:0 0 gic_it_set_prio:266 prio: writing 0x1 to 0x11f00428
I/TC: Initialized
D/TC:0 0 init_primary_helper:1060 Primary CPU switching to normal world boot


U-Boot 2018.07-g6c35bd5 (Feb 25 2019 - 22:43:49 +0100)

```

One can note Qemu-Armv7 TF-A outputs its console through UART0 instance. This is built-in the firmware.

#### Some fix proposals
We could use the overlay (`update_external_dt()` in `generic_boot.c`) to add a `secure-chosen` node if none found, and store a `stdout-path` property `/pl011@09040000`.
This would need a platform hook in `update_external_dt()`. 

We could disable DT support only for the console. Would need a configuration directive like `CFG_IGNORE_DT_STDOUT_PATH`.

We could preserve the current sequence (early init on UART1 then probed from DT console, here UART0) and have the `qemu_virt` platform registering again the console on UART1 using a `service_init` or `driver_init` initcall.
I am in favor for this one has no extra hooks and configs. Console toggles a bit but that's definitively not an issue on Qemu, and may show other platforms several console initialization examples. Alternatively, it can be confusing :)

**Comment by jforissier** [comment]
> The issue is due to the console being probed from the external DT which provides only the non-secure console information.

What version of QEMU are you using? You are probably missing this patch: https://github.com/qemu/qemu/commit/fb23d69 which is included in v3.1.0.

```bash
root@Vexpress:/ cat /sys/firmware/devicetree/base/secure-chosen/stdout-path; echo
/pl011@9040000
```

**Comment by etienne-lms** [comment]
ok, thanks @jforissier, much better.
closing issue.
